src/coreComponents/functions/MultivariableTableFunctionKernels.hpp

```diff
@@ -0,0 +1,321 @@
+   * @param[out] axisLow left coordinate of target axis interval
+   * @param[out] axisMult weight of the right coordinate of target axis interval
+   * @return integer target axis interval index
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  integer
+  getAxisIntervalIndexLowMult( real64 const axisCoordinate,
+                               real64 const axisMin,
+                               real64 const axisMax,
+                               real64 const axisStep,
+                               real64 const axisStepInv,
+                               integer const axisPoints,
+                               real64 & axisLow,
+                               real64 & axisMult ) const
+  {
+    integer axisIntervalIndex = integer((axisCoordinate - axisMin) * axisStepInv );
+
+    // check that axisindex is within interpolation interval: valid axisindex is between 0 and (axisPoints - 2),
+    // since there are axisPoints-1 intervals along the axis
+
+    if( axisIntervalIndex < 0 )
+    {
+      axisIntervalIndex = 0;
+      if( axisCoordinate < axisMin )
+      {
+        printf( "Interpolation warning: axis is out of limits (%lf; %lf) with value %lf, extrapolation is applied\n", axisMin, axisMax, axisCoordinate );
+      }
+    }
+    else if( axisIntervalIndex > (axisPoints - 2))
+    {
+      axisIntervalIndex = axisPoints - 2;
+      if( axisCoordinate > axisMax )
+      {
+        printf( "Interpolation warning: axis is out of limits (%lf; %lf) with value %lf, extrapolation is applied\n", axisMin, axisMax, axisCoordinate );
+      }
+    }
+
+    axisLow = axisIntervalIndex * axisStep + axisMin;
+    axisMult = (axisCoordinate - axisLow) * axisStepInv;
+    return axisIntervalIndex;
+  }
+
+
+  /**
+   * @brief interpolate all operator values and derivatives at a given point
+   *
+   * @param[in] axisCoordinates coordinates of a point
+   * @param[in] hypercubeData data of target hypercube
+   * @param[in] axisLows array of left coordinates of target axis intervals
+   * @param[in] axisMults array of weights of right coordinates of target axis intervals
+   * @param[in] axisStepInvs array of inversions of axis steps
+   * @param[out] values interpolated operator values
+   * @param[out] derivatives derivatives of interpolated operators
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  void
+  interpolatePointWithDerivatives( real64 const * const axisCoordinates,
+                                   real64 const * const hypercubeData,
+                                   real64 const * const axisLows,
+                                   real64 const * const axisMults,
+                                   real64 const * const axisStepInvs,
+                                   real64 *values,
+                                   real64 *derivatives ) const
+  {
+    static constexpr integer NUM_VERTS = 1 << NUM_DIMS;
+    integer pwr = NUM_VERTS / 2;   // distance between high and low values
+    real64 workspace[(2 * NUM_VERTS - 1) * NUM_OPS];
```

Comment:
I`ve always thought that C++ 1D arrays are always faster than 2D ones, but it seems that is not true anymore, right?
In any case, indexing looks better with 2D arrays for sure.

Reply:
In the case of stack-based C-arrays with compile-time bounds, they always have identical memory representation and codegen (example: https://godbolt.org/z/x9T1ve68G). When arrays are dynamic and heap-allocated, you are of course right and we try to avoid things like `double **` or `vector<vector<T>>` (this is where our `Array<T,N>` abstraction comes in, as it provides multidimensional view of a single contiguous memory allocation).

Reply:
Thanks, @klevzoff, very clear and evident!